gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/extractor/HiveConvertExtractor.java

```diff
@@ -39,54 +45,74 @@
 
-  private static final String OPTIONAL_HIVE_AVRO_SCHEMA_PROVIDER_CLASS_KEY = "hive.unit.avroSchemaProvider.class";
-  private static final String DEFAULT_HIVE_AVRO_SCHEMA_PROVIDER_CLASS = HdfsBasedSchemaProvider.class.getName();
+  private List<QueryBasedHiveConversionEntity> conversionEntities = Lists.newArrayList();
 
-  /**
-   * Even though each extractor only processes one {@link HiveRegistrationUnit}, We use a list with one {@link HiveRegistrationUnit}
-   * so that {@link #readRecord(QueryBasedHiveConversionEntity)} knows when to stop.
-   */
-  private final List<HiveRegistrationUnit> hiveUnits;
-  private final HiveAvroSchemaProvider hiveAvroSchemaProvider;
+  public HiveConvertExtractor(WorkUnitState state, FileSystem fs) throws IOException, TException, HiveException {
+
+    HiveMetastoreClientPool pool =
+        HiveMetastoreClientPool.get(state.getJobState().getProperties(),
+            Optional.of(state.getJobState().getProp(HiveDatasetFinder.HIVE_METASTORE_URI_KEY)));
 
-  public HiveConvertExtractor(WorkUnitState state, FileSystem fs) {
-    HiveRegistrationUnit hiveUnit =
-        HiveSource.GENERICS_AWARE_GSON.fromJson(state.getProp(HiveSource.HIVE_UNIT_SERIALIZED_KEY),
-            HiveRegistrationUnit.class);
+    SerializableHiveTable hiveTable = HiveSourceUtils.deserializeTable(state);
 
-    this.hiveUnits = Lists.newArrayList(hiveUnit);
-    this.hiveAvroSchemaProvider =
-        GobblinConstructorUtils.invokeConstructor(HiveAvroSchemaProvider.class,
-            state.getProp(OPTIONAL_HIVE_AVRO_SCHEMA_PROVIDER_CLASS_KEY, DEFAULT_HIVE_AVRO_SCHEMA_PROVIDER_CLASS), fs);
+    Table table = pool.getClient().get().getTable(hiveTable.getDbName(), hiveTable.getTableName());
+
+    SchemaAwareHiveTable schemaAwareHiveTable =
+        new SchemaAwareHiveTable(table, AvroSchemaManager.getSchemaFromUrl(hiveTable.getSchemaUrl(), fs));
+
+    SchemaAwareHivePartition schemaAwareHivePartition = null;
+
+    if (HiveSourceUtils.hasPartition(state)) {
+
+      SerializableHivePartition hivePartition = HiveSourceUtils.deserializePartition(state);
+
+      Partition partition =
+          pool.getClient().get()
+              .getPartition(hiveTable.getTableName(), hiveTable.getDbName(), hivePartition.getPartitionName());
+      schemaAwareHivePartition =
+          new SchemaAwareHivePartition(table, partition, AvroSchemaManager.getSchemaFromUrl(
+              hivePartition.getSchemaUrl(), fs));
+    }
+
+    this.conversionEntities.add(new QueryBasedHiveConversionEntity(schemaAwareHiveTable, Optional
+        .fromNullable(schemaAwareHivePartition)));
 
   }
 
   @Override
   public Schema getSchema() throws IOException {
-    return this.hiveAvroSchemaProvider.getSchema(this.hiveUnits.get(0));
-  }
+    if (this.conversionEntities.isEmpty()) {
+      return null;
+    }
 
+    QueryBasedHiveConversionEntity conversionEntity = this.conversionEntities.get(0);
+    if (conversionEntity.getHivePartition().isPresent()) {
```

Comment:
This will ensure that the older as well as newer partitions both are consistent, and we can handle evolution when we add support for it